File: runtimes/python/src/wrapper_runtime/transport/streamable_http.py
"""T064: `serve --mode streamable-http` for the Python runtime (FR-015, FR-016).

Runs the generated package as a hosted service using the official MCP Python SDK's Streamable
HTTP transport (never the deprecated HTTP+SSE transport — Constitution Principle I). The SDK
import is deferred so this module imports without the SDK installed.
"""
from __future__ import annotations

from ..manifest import LoadedPackage
from ..server import register_tools


def serve_streamable_http(pkg: LoadedPackage, port: int, base_url: str) -> int:
    """Start a Streamable HTTP MCP server for the package."""
    try:
        from mcp.server.fastmcp import FastMCP  # type: ignore

        server = FastMCP(f"wrapper-{pkg.manifest['webappTargetId']}")
        register_tools(pkg, server, base_url)
        # FastMCP exposes a streamable-http runner; select it explicitly (not SSE).
        server.run(transport="streamable-http", port=port)
        return 0
    except Exception as exc:  # noqa: BLE001
        import sys

        sys.stderr.write(f"Streamable HTTP transport unavailable ({exc}).\n")
        return 1

File: runtimes/node/src/transport/streamable-http.ts
// T063: `serve --mode streamable-http` for the Node runtime (FR-015, FR-016).
// Runs the generated package as a hosted service using the MCP Streamable HTTP transport
// (never the deprecated HTTP+SSE transport — Constitution Principle I). The SDK transport is
// imported lazily so this module builds without the SDK present.

import type { LoadedPackage } from "../manifest.js";
import { registerTools, type ToolDispatchContext } from "../server.js";

export interface StreamableHttpOptions {
  port: number;
  redirectUri?: string;
  baseUrl: string;
}

/** Start a Streamable HTTP MCP server for the package. Returns a stop() handle. */
export async function serveStreamableHttp(
  pkg: LoadedPackage,
  opts: StreamableHttpOptions,
): Promise<{ port: number; stop: () => Promise<void> }> {
  const ctx: ToolDispatchContext = { baseUrl: opts.baseUrl };
  const { register } = await registerTools(pkg, ctx);

  const sdk = await import("@modelcontextprotocol/sdk/server/mcp.js");
  const httpMod = await import("@modelcontextprotocol/sdk/server/streamableHttp.js");

  const McpServer = (sdk as unknown as { McpServer: new (o: unknown) => { connect(t: unknown): Promise<void> } & Parameters<typeof register>[0] }).McpServer;
  const server = new McpServer({
    name: `wrapper-${pkg.manifest.webappTargetId}`,
    version: pkg.manifest.runtimeVersion,
  });
  register(server);

  const TransportCtor = (httpMod as unknown as {
    StreamableHTTPServerTransport: new (o: unknown) => { close?: () => Promise<void> };
  }).StreamableHTTPServerTransport;
  const transport = new TransportCtor({ port: opts.port });
  await server.connect(transport);

  return {
    port: opts.port,
    stop: async () => {
      await transport.close?.();
    },
  };
}
